app/src/main/java/de/danoeh/antennapod/activity/SingleFileAddActivity.java

```diff
@@ -0,0 +1,220 @@
+        if (permission != PackageManager.PERMISSION_GRANTED) {
+            requestPermission();
+        }
+
+        try {
+            startActivityForResult(intentPickAction, CHOOSE_SINGLE_FILE);
+        } catch (ActivityNotFoundException e) {
+            Log.e(TAG, "No activity found. Should never happen...");
+        }
+    }
+
+    /**
+      * Gets the path of the file chosen with chooseFileToImport()
+      */
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        //super.onActivityResult(requestCode, resultCode, data);
+        if (resultCode == RESULT_OK && requestCode == CHOOSE_SINGLE_FILE) {
+            Uri uri = data.getData();
+            if(uri != null && uri.toString().startsWith("/")) {
+                uri = Uri.parse("file://" + uri.toString());
+            }
+            boolean ret = importUri(uri);
+            if (ret == false) {
+                Toast.makeText(this, "Could not import", Toast.LENGTH_LONG).show();
+            }
+
+            //go back to main
+            Intent intent = new Intent(this, MainActivity.class);
+            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP
+                    | Intent.FLAG_ACTIVITY_NEW_TASK);
+            startActivity(intent);
+        } else {
+            Log.d(TAG, "result not ok or code not file: " + resultCode + "," + requestCode);
+        }
+    }
+
+    private void requestPermission() {
+        String[] permissions = { android.Manifest.permission.READ_EXTERNAL_STORAGE };
+        ActivityCompat.requestPermissions(this, permissions, PERMISSION_REQUEST_READ_EXTERNAL_STORAGE);
+    }
+
+    boolean importUri(@Nullable Uri uri) {
+        if(uri == null) {
+            new MaterialDialog.Builder(this)
+                    .content(R.string.opml_import_error_no_file)
+                    .positiveText(android.R.string.ok)
+                    .show();
+            return false;
+        }
+        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
+                uri.toString().contains(Environment.getExternalStorageDirectory().toString())) {
+            int permission = ActivityCompat.checkSelfPermission(this, android.Manifest.permission.READ_EXTERNAL_STORAGE);
+            if (permission != PackageManager.PERMISSION_GRANTED) {
+                return false;
+            }
+            startImport(uri);
+            return true;
+        }
+        return false;
+    }
+
+    private long getFileDuration(File f) {
+        MediaMetadataRetriever mediaMetadataRetriever = new MediaMetadataRetriever();
+        mediaMetadataRetriever.setDataSource(f.getAbsolutePath());
+        String durationStr = mediaMetadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION);
+        return Long.parseLong(durationStr);
+    }
+
+    /** Starts the import process. */
+    private void startImport(Uri uri) {
```

Comment:
Please move this to a new class, for example `core/src/main/java/de/danoeh/antennapod/core/feed/LocalFeedUpdater.java`

Reply:
Should `LocalFeedUpdater.startImport` also receive a reference to `AddLocalFilesActivity` as a `Context` (whatever that is)?
Because `DBTasks.updateFeed` requires one.

Reply:
> Because DBTasks.updateFeed requires one.

yes :)